### DISCUSSION TITLE
Where/how is MATCHER_HOOK used?

### DISCUSSION THREAD

**Discussion by vsajip** [congo-cc/congo-parser-generator#80] [Q&A] [answered]
https://github.com/congo-cc/congo-parser-generator/blob/feb0e8bf2b6ba9b3970b855874a2283c173feefb/src/templates/java/Lexer.java.ftl#L56
and
https://github.com/congo-cc/congo-parser-generator/blob/feb0e8bf2b6ba9b3970b855874a2283c173feefb/src/templates/java/Lexer.java.ftl#L292-L293
are the only references to `MATCHER_HOOK`. Given that it's a `private static`, where and how is it meant to be set?

**Comment by revusky** [comment]
The `MATCHER_HOOK` thing was added by  John Bradley and, as best I understand, allows an optional optimization that he was (I assume still is) interested in. I have to admit that I haven't even really looked at it.

I think the bottom line for now is that, in terms of bringing the Python/CSharp templates up to date, you could perfectly well, just ignore that, i.e. just assume for now that it's always null.

**Comment by vsajip** [comment]
Wouldn't it be cleaner to accommodate this by making `getMatchInfo()` non-static and `protected` and taking a `LexicalState` initial argument? Then @adMartem could just override `getmatchInfo()` in a parser injection where needed.

**Comment by revusky** [comment]
> Wouldn't it be cleaner to accommodate this by making `getMatchInfo()` non-static and `protected` and taking a `LexicalState` initial argument? Then @adMartem could just override `getmatchInfo()` in a parser injection where needed.

Well, I have to admit that I haven't really looked at what @adMartem  was doing with that. So, maybe you're right. I'm not sure.

As for `getMatchInfo()`, well, I still have the (possibly vague, speculative) notion that being able to call it from a static context could be useful. Though, admittedly, not that I am currently extracting any value from that, but...  It takes as an argument the `NFAFunction[]` array, which is basically what the current lexical state amounts to -- well, the actual state of the world in terms of lexing is that plus the set of active tokens, which is also passed in.

But, as I said, I don't think there is such a strong need (at least at the moment) to worry about whatever extra stuff @adMartem has been doing with MATCHER_HOOK -- I mean, supporting that in non-Java languages. It's quite possible that there will always be a few extra things that work only in Java. The problem is that we have very BIG key things that don't really work, namely code injection. Well, there is the transpilation approach, yes, but it doesn't seem to be so robust. It would be nice to be able to simply inject code in Python itself or in CSharp itself... in a way that is analogous to how it worksfor Java. But I guess I'll have to buckle down and finally get that working. A big piece has been done for a while, which is that CongoCC itself actually does have the ability to parse both C# and Python internally. And there is the ability for all the nodes to be in one uber-tree. That was the big technical accomplishment from, I dunno, like 10 months ago, but I somehow got distracted by other things and...

**Comment by adMartem** [comment, marked as answer]
I added the functional interface the way I did because I didn't want to perturb what Jon had just finished doing with the Lexer refactoring and I wanted to "hook" into the lexer code in what I thought would be the most efficient way (for Java/Hotspot) since it is such a critical path. I didn't expect the hook to be used in any language except Java, since I only use Java and, if it made sense in any other languages, the DFA optimization I use it for should be done at lexer generation time anyway.  So just leave it out of the other language if this is still a question.  I had referred to this on Slack a while back, but I suspect at that time @vsajip didn't have context in mind to make sense out of my comment ;-)
